File: algorithms/python/144-BinaryTreePreorderTraversal/main.py
#
# https://leetcode.com/problems/binary-tree-preorder-traversal/
# Easy
# Topics: Array, Tree, Depth First Search, Binary Tree
#
# Pre-order traversal is a traversal type used in Depth First Search (DFS). We can do this iteratively or recuresively.
# If done iteratively, it is done using a stack

# Time Complexity: O(n)
# Space Complexity: O(n)
#

from typing import List
from typing import Optional
from common.Tree import Tree
from common.TreeNode import TreeNode

import unittest

RECURSIVE = False

if RECURSIVE:
    class Solution:
        def dfs(self, root: Optional[TreeNode], results: List[int]):
            if not root:
                return
            results.append(root.val)
            if root.left:
                self.dfs(root.left, results)
            if root.right:
                self.dfs(root.right, results)

        def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
            results: List[int] = []
            self.dfs(root, results)
            return results
else:
    class Solution:
        def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
            results: List[int] = []
            stack: List[TreeNode] = []

            stack.append(root)

            while stack:
                node = stack.pop()
                if node:
                    results.append(node.val)
                    stack.append(node.right)
                    stack.append(node.left)

            return results

class TestSolution(unittest.TestCase):
    def setUp(self):
        self.soln = Solution()

    def test_case0(self):
        test_case0 = Tree([1, None, 2, 3])
        self.assertCountEqual(self.soln.preorderTraversal(test_case0.root), [1, 2, 3])

    def test_case1(self):
        test_case1 = Tree([])
        self.assertCountEqual(self.soln.preorderTraversal(test_case1.root), [])

    def test_case2(self):
        test_case2 = Tree([1])
        self.assertCountEqual(self.soln.preorderTraversal(test_case2.root), [1])

    def test_case3(self):
        test_case3 = Tree([1, 2, 3, 4, 5, 6, 7])
        self.assertCountEqual(self.soln.preorderTraversal(test_case3.root), [1, 2, 3, 4, 5, 6, 7])

File: algorithms/cpp/144-BinaryTreePreorderTraversal/main.cpp
//
// Created by Benjamin Lee on 1/8/23.
// https://leetcode.com/problems/binary-tree-preorder-traversal/
// Easy
// Topics:
//

/*
 * This translates to a Depth First Search (DFS).
 *
 * For a refresher, Depth First Search goes down the children. There are 3 types of traversal:
 * Pre-Order: root, left, right
 * In-Order: left, root, right
 * Post-Order: right, root, left
 *
 * We can approach this two ways:
 *
 * Recursion
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * Iterative
 * Iterative approach uses a stack
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

#define CATCH_CONFIG_MAIN
#include <catch2/catch_all.hpp>

#include <stack>
#include "../include/BinarySearchTree.h"

class Solution {
public:
#ifdef RECURSIVE
    void dfs(TreeNode *root, std::vector<int>& results) {
        if (!root) return;

        results.push_back(root->val);

        if (root->left) {
            dfs(root->left, results);
        }

        if (root->right) {
            dfs(root->right, results);
        }
    }

    std::vector<int> preorderTraversal(TreeNode *root) {
        std::vector<int> results;

        dfs(root, results);

        return results;
    }
#else
    std::vector<int> preorderTraversal(TreeNode *root) {
        if (!root) return {};

        std::vector<int> results;
        std::stack<TreeNode *> stack;

        stack.push(root);

        while (stack.size()) {
            auto node = stack.top();
            stack.pop();

            if (node) {
                results.push_back(node->val);
                stack.push(node->right);
                stack.push(node->left);
            }
        }

        return results;
    }
#endif /* RECURSIVE */
};

TEST_CASE("BinaryTreePreorderTraversal", "algorithm") {
    Solution soln;
    Tree testCase0{1, std::numeric_limits<int>::min(), 2, 3};
    Tree testCase1{};
    Tree testCase2{1};
    Tree testCase3{1, 2, 3, 4, 5, 6, 7};

    CHECK(soln.preorderTraversal(testCase0.root) == std::vector<int>{1, 2, 3});
    CHECK(soln.preorderTraversal(testCase1.root) == std::vector<int>{});
    CHECK(soln.preorderTraversal(testCase2.root) == std::vector<int>{1});
    CHECK(soln.preorderTraversal(testCase3.root) == std::vector<int>{1, 2, 4, 5, 3, 6, 7});
}
